### ISSUE TITLE
WSL2 tensorflow install not working

### ISSUE THREAD

**Issue by Syndicateeee** [tensorflow/tensorflow#62095]
### Issue type

Build/Install

### Have you reproduced the bug with TensorFlow Nightly?

No

### Source

source

### TensorFlow version

v2.14.0-rc1-21-g4dacf3f368e 2.14.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04.1 on WSL2

### Python version

3.10.12

### CUDA/cuDNN version

CUDA: 12.2 cudnn: ??

### GPU model and memory

rtx 3090 24gb 

### Current behavior?

Im trying to use tensorflow 2 on wsl2. I did the installation as in the tf docs.
After Installing tensorflow on wsl2 and running a test script as stated in the relevant code.

I get this some errors.

I have tried training a simple model with resnet which, at first, didnt work because cudnn was missing. After manually installing cudnn it worked but it feels so sketchy given that it should work without having to install it manually.

Also when I train my model (i dont know if this is normal behaviour) only my vram is used, the compute power of my gpu seems to not be used at all when i look at my taskmanager.

### Standalone code to reproduce the issue

```shell
import tensorflow as tf
tf.test.is_built_with_cuda()
```

### Relevant log output

```shell
2023-10-11 22:19:14.589975: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-10-11 22:19:14.590010: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-10-11 22:19:14.590032: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-10-11 22:19:14.593224: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.
/usr/lib/python3/dist-packages/scipy/__init__.py:146: UserWarning: A NumPy version >=1.17.3 and <1.25.0 is required for this version of SciPy (detected version 1.26.0
  warnings.warn(f"A NumPy version >={np_minversion} and <{np_maxversion}"
```

**Comment by SuryanarayanaY** [comment]
Hi @Syndicateeee ,

Could you please confirm the installation command you have used ? Is it `python3 -m pip install tensorflow[and-cuda]` ??

Also please note that above command will fetch the required cuda,cudnn libraries automatically. So please don't install any cuda libraries separately as it seems duplicating the registries. Only you need to install suitable CUDA driver manually and then all required CUDA toolkit shall be imported along with tensorflow if used above command.

Thank you.

**Comment by Syndicateeee** [comment]
@SuryanarayanaY yes this is the exact command, i also tried it with docker desktop on windows and docker in wsl2. windows wont boot at all which is fine probably misconfigured and wsl2 also yields the same error.

**Comment by SuryanarayanaY** [comment]
@Syndicateeee ,

I hope you have deleted any CUDA libraries that is installed manually except those bundled with tensorflow[and-cuda].Please confirm. Because multiple libraries may cause ambiguity.

I expect nvidia driver has been installed and running.Please share `nvidia-smi` command output.

Please share the environment details using `pip list` .

I am attaching a relevant issues here #57663 for verifying.

**Comment by Syndicateeee** [comment]
@SuryanarayanaY Sorry for the long wait, I installed ubuntu on a seperate drive for my system, to see if its a windows/wsl2 error. Its a complete fresh install, the first thing i did is install the nvidia driver and then run the install of tensorflow (same command as previosly) but to no avail, the error is exactly the same. I find it weird that it states "where one has already been registered" altho there is only the cuda packages which get downloaded with tensorflow.

nvidia-smi: 

+---------------------------------------------------------------------------------------+
| NVIDIA-SMI 535.113.01             Driver Version: 535.113.01   CUDA Version: 12.2     |
|-----------------------------------------+----------------------+----------------------+
|=========================================+======================+======================|
|   0  NVIDIA GeForce RTX 3090        Off | 00000000:06:00.0  On |                  N/A |
|  0%   42C    P8              36W / 420W |    388MiB / 24576MiB |      4%      Default |



pip list: 

Package                      Version
---------------------------- ----------------
absl-py                      2.0.0
apturl                       0.5.2
astunparse                   1.6.3
bcrypt                       3.2.0
blinker                      1.4
Brlapi                       0.8.3
cachetools                   5.3.1
certifi                      2020.6.20
chardet                      4.0.0
click                        8.0.3
colorama                     0.4.4
command-not-found            0.3
cryptography                 3.4.8
cupshelpers                  1.0
dbus-python                  1.2.18
defer                        1.0.6
distro                       1.7.0
distro-info                  1.1+ubuntu0.1
duplicity                    0.8.21
fasteners                    0.14.1
flatbuffers                  23.5.26
future                       0.18.2
gast                         0.5.4
google-auth                  2.23.3
google-auth-oauthlib         1.0.0
google-pasta                 0.2.0
grpcio                       1.59.0
h5py                         3.10.0
httplib2                     0.20.2
idna                         3.3
importlib-metadata           4.6.4
jeepney                      0.7.1
keras                        2.14.0
keyring                      23.5.0
language-selector            0.1
launchpadlib                 1.10.16
lazr.restfulclient           0.14.4
lazr.uri                     1.0.6
libclang                     16.0.6
lockfile                     0.12.2
louis                        3.20.0
macaroonbakery               1.3.1
Mako                         1.1.3
Markdown                     3.5
MarkupSafe                   2.1.3
ml-dtypes                    0.2.0
monotonic                    1.6
more-itertools               8.10.0
netifaces                    0.11.0
numpy                        1.26.0
nvidia-cublas-cu11           11.11.3.6
nvidia-cuda-cupti-cu11       11.8.87
nvidia-cuda-nvcc-cu11        11.8.89
nvidia-cuda-runtime-cu11     11.8.89
nvidia-cudnn-cu11            8.7.0.84
nvidia-cufft-cu11            10.9.0.58
nvidia-curand-cu11           10.3.0.86
nvidia-cusolver-cu11         11.4.1.48
nvidia-cusparse-cu11         11.7.5.86
nvidia-nccl-cu11             2.16.5
oauthlib                     3.2.0
olefile                      0.46
opt-einsum                   3.3.0
packaging                    23.2
paramiko                     2.9.3
pexpect                      4.8.0
Pillow                       9.0.1
pip                          23.2.1
protobuf                     4.24.4
ptyprocess                   0.7.0
pyasn1                       0.5.0
pyasn1-modules               0.3.0
pycairo                      1.20.1
pycups                       2.0.1
PyGObject                    3.42.1
PyJWT                        2.3.0
pymacaroons                  0.13.0
PyNaCl                       1.5.0
pyparsing                    2.4.7
pyRFC3339                    1.1
python-apt                   2.4.0+ubuntu2
python-dateutil              2.8.1
python-debian                0.1.43+ubuntu1.1
pytz                         2022.1
pyxdg                        0.27
PyYAML                       5.4.1
reportlab                    3.6.8
requests                     2.25.1
requests-oauthlib            1.3.1
rsa                          4.9
screen-resolution-extra      0.0.0
SecretStorage                3.3.1
setuptools                   59.6.0
six                          1.16.0
systemd-python               234
tensorboard                  2.14.1
tensorboard-data-server      0.7.1
tensorflow                   2.14.0
tensorflow-estimator         2.14.0
tensorflow-io-gcs-filesystem 0.34.0
tensorrt                     8.5.3.1
termcolor                    2.3.0
typing_extensions            4.8.0
ubuntu-advantage-tools       8001
ubuntu-drivers-common        0.0.0
ufw                          0.36.1
unattended-upgrades          0.1
urllib3                      1.26.5
usb-creator                  0.3.7
wadllib                      1.3.6
Werkzeug                     3.0.0
wheel                        0.37.1
wrapt                        1.14.1
xdg                          5
xkit                         0.0.0
zipp                         1.0.0

The full error when i run "import tensorflow":

import tensorflow
2023-10-13 18:44:23.475339: I tensorflow/core/util/port.cc:111] oneDNN custom operations are on. You may see slightly different numerical results due to floating-point round-off errors from different computation orders. To turn them off, set the environment variable `TF_ENABLE_ONEDNN_OPTS=0`.
2023-10-13 18:44:23.496059: E tensorflow/compiler/xla/stream_executor/cuda/cuda_dnn.cc:9342] Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered
2023-10-13 18:44:23.496077: E tensorflow/compiler/xla/stream_executor/cuda/cuda_fft.cc:609] Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered
2023-10-13 18:44:23.496091: E tensorflow/compiler/xla/stream_executor/cuda/cuda_blas.cc:1518] Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered
2023-10-13 18:44:23.499851: I tensorflow/core/platform/cpu_feature_guard.cc:182] This TensorFlow binary is optimized to use available CPU instructions in performance-critical operations.
To enable the following instructions: AVX2 AVX_VNNI FMA, in other operations, rebuild TensorFlow with the appropriate compiler flags.

**Comment by Syndicateeee** [comment]
I managed to resolve the issue. Note that im not sure what the issue was in the end. This worked for tensorflow 2.14
My steps to solving it:
1. installed Ubuntu 20.04 LTS
2. disable nouveau driver [https://askubuntu.com/a/951892](url)
3. install cudaToolkit 11.8 using the .run local installer [https://developer.nvidia.com/cuda-11-8-0-download-archive?target_os=Linux&target_arch=x86_64&Distribution=Ubuntu&target_version=20.04&target_type=runfile_local](url)
4. add cuda path to PATH and LD_LIBRARY_PATH [https://forums.developer.nvidia.com/t/path-ld-library-path/48080](url)
5. install cudnn 8.6 using the installation guide [https://docs.nvidia.com/deeplearning/cudnn/install-guide/index.html#installlinux-deb](url)
6. check if cuda is working `nvcc --version`
7. check if cudnn is working as stated in the installer. 
7.1. if you get an error about FreeImage not being setup correctly [https://forums.developer.nvidia.com/t/freeimage-is-not-set-up-correctly-please-ensure-freeimae-is-set-up-correctly/66950/2?u=tschm300](url)
7.2. If you get an error stating somethin about "compute_35" go to cudnn_samples_v8 folder edit the samples common file `sudo nano samples_common.mk ` and at the end where it says `SMS ?=` remove 35 and save.
7.3. if your test passes you have installed cuda and cudnn
8. install anaconda [https://docs.anaconda.com/free/anaconda/install/linux/](url)
9. create anacona environment using `conda create --name tensorflow_001 -c conda-forge python=3.11
`
10. install tensorflow using `python3 -m pip install tensorflow[and-cuda]`
11. done (note: some infos may be still there)

**Comment by FaisalAlj** [comment]
Hello, 

I'm experiencing the same issue and  I have  already posted in details what I'm dealing with in this comment [https://github.com/tensorflow/tensorflow/issues/62075#issuecomment-1766011448](url)

Error:
Unable to register cuDNN factory: Attempting to register factory for plugin cuDNN when one has already been registered.
Unable to register cuFFT factory: Attempting to register factory for plugin cuFFT when one has already been registered.
Unable to register cuBLAS factory: Attempting to register factory for plugin cuBLAS when one has already been registered.

OS: Windows 11, WSL 2, Ubuntu 22.04.3 LTS

**Comment by Syndicateeee** [comment]
@FaisalAlj 
1. tensorflow uses cuDNN 8.6
2. for me installing the 11.8 toolkit wasnt enough, i also needed a NVIDIA driver with cuda 11.8 which is 520.61.05
3. this driver only worked on Ubuntu 20.04 supposedly because of the newer Ubuntu kernel in 22.04
4. I gave up on WSL2, went to a close by electronics store, bought a 500gig nvme and installed ubuntu on there
5. If you ensist on using wsl2 please give my guide a shot [https://github.com/tensorflow/tensorflow/issues/62095#issuecomment-1763366758](url)
You can setup ubuntu 20.04 on wsl2 using `wsl --install -d Ubuntu-20.04` also make sure ur using wsl2 not wsl so after u install `wsl --set-version Ubuntu-20.04 2` not sure if you can install the driver on wsl tho might look for the same nvidia driver for your windows host. When typing `nvidia-smi` in ur terminal it should say CUDA: 11.8.

**Comment by DTraitor** [comment]
I was trying to slove this issue for a while now and nothing seemed to work. I've tried using Windows (tensorflow no longer supports it), VirtualBox (doesn't provide direct access to GPU) and WSL2. 
I've tried installing CUDA toolkit, WSL, conda, packages, cudnn, everything, nothing helped. And you know what did? 
USING, F*CKING **PYTHON 3.11** INSTEAD OF 3.12!

**Comment by DTraitor** [comment]
Well, looks like I solved it. I've tried everything to make CUDA work, from installing older tensorflow-gpu to VirtualBox and WSL2. First of all, downgrade to 2.16.1, it shows more info on whats wrong. Next step search for an env variable that controls TF logs (something like "TF_CPP_MAX_LOG_LEVEL"), set it to higher level so it shows you what is exactly going wrong (which libraries are there and which are not). Then uninstall all CUDA and cudnn stuff from WSL and install CUDA 12.3 and cudnn 8.9. Exactly these version. Installing them is a bit tricky because you have to specify them in apt-get and Nvidia made installing cudnn8 pain in the ass after they released cudnn9. So after installing those in my WSL2, when I compile my model it uses GPU memory, alas there are some NUMA related errors which I wasn't able to fix but training became much faster either way (40s->5s per epoch)
